File: app/src/main/java/com/donald/musictheoryapp/Question/TextInputQuestion.java
package com.donald.musictheoryapp.Question;

import com.donald.musictheoryapp.Utils.JSONArrayUtil;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Objects;

// TODO: 1 ANSWER PER TEXT INPUT? OR NOT?

public class TextInputQuestion extends Question
{
    // TODO: CHANGE TO INT
    public enum InputType
    {
        Text,
        Number
    }

    public static class Answer implements Question.Answer
    {
        public String userAnswer;
        public String correctAnswer;

        public Answer(String userAnswer, String correctAnswer)
        {
            this.userAnswer = userAnswer;
            this.correctAnswer = correctAnswer;
        }

        public JSONObject toJSON() throws JSONException
        {
            JSONObject object = new JSONObject();
            object.put("user_answer", userAnswer == null ? JSONObject.NULL : userAnswer);
            object.put("correct_answer", correctAnswer);
            return object;
        }

        public static Answer fromJSON(JSONObject object) throws JSONException
        {
            return new Answer(
                object.isNull("user_answer") ? null : object.getString("user_answer"),
                object.getString("correct_answer")
            );
        }

        @Override
        public boolean correct()
        {
            return userAnswer != null && userAnswer.equals(correctAnswer);
        }
    }

    public InputType inputType;
    public Answer[] answers;

    public TextInputQuestion() {}

    @Override
    public int points()
    {
        int points = 0;
        for(Answer answer : answers)
        {
            if(answer.correct()) points++;
        }
        return points;
    }

    @Override
    public int maxPoints()
    {
        return answers.length;
    }

    @Override
    public void acceptVisitor(QuestionVisitor visitor) { visitor.visit(this); }

    public static TextInputQuestion fromJSON(JSONObject object, QuestionGroup group) throws JSONException
    {
        TextInputQuestion question = new TextInputQuestion();
        question.number = object.getInt("number");
        question.group = group;
        question.descriptions = JSONArrayUtil.descriptions(object);
        question.inputType = InputType.values()[object.getInt("input_type")];
        JSONArray answerJSONArray = object.getJSONArray("answers");
        question.answers = new Answer[answerJSONArray.length()];
        for(int i = 0; i < question.answers.length; i++)
        {
            question.answers[i] = Answer.fromJSON(answerJSONArray.getJSONObject(i));
        }
        return question;
    }
}

File: app/src/main/java/com/donald/musictheoryapp/question/TextInputQuestion.kt
package com.donald.musictheoryapp.question

import com.donald.musictheoryapp.util.getDescriptions
import com.donald.musictheoryapp.util.getInputHint
import com.donald.musictheoryapp.util.getInputType
import org.json.JSONArray
import org.json.JSONObject

// TODO: 1 ANSWER PER TEXT INPUT? OR NOT?
class TextInputQuestion(
    number: Int,
    descriptions: List<Description>,
    inputHint: String?,
    val inputType: InputType,
    val answers: List<Answer>
) : ChildQuestion(number, descriptions, inputHint) {

    override val points: Int
        get() = answers.count { it.correct }
    override val maxPoints = answers.size
    override val correctAnswerCount = answers.sumOf { it.correctAnswers.size }

    override fun acceptVisitor(visitor: QuestionVisitor) {
        visitor.visit(this)
    }

    override fun copyNew(): TextInputQuestion {
        return TextInputQuestion(
            number,
            List(descriptions.size) { i -> descriptions[i].copy() },
            inputHint,
            inputType,
            List(answers.size) { i -> answers[i].copyNew() }
        )
    }

    override fun toPartialJson(): JSONObject {
        val jsonObject = JSONObject()
        jsonObject.put(QUESTION_TYPE, Type.TEXT_INPUT.ordinal)
        jsonObject.put(INPUT_TYPE, inputType.ordinal)
        val array = JSONArray()
        for (answer in answers) {
            array.put(answer.toJson())
        }
        jsonObject.put("answers", array)
        return jsonObject
    }

    companion object {

        fun fromJson(jsonObject: JSONObject): TextInputQuestion {
            return TextInputQuestion(
                number = jsonObject.getInt("number"),
                descriptions = jsonObject.getDescriptions(),
                inputHint = jsonObject.getInputHint(),
                inputType = jsonObject.getInputType(),
                answers = jsonObject.getJSONArray("answers").run {
                    List(this.length()) { index ->
                        Answer.fromJson(this.getJSONObject(index))
                    }
                }
            )
        }

    }

    enum class InputType {
        Text, Number
    }

    class Answer(var userAnswer: String?, val correctAnswers: List<String>) : ChildQuestion.Answer {

        override val correct: Boolean
            get() = correctAnswers.any { it == userAnswer }

        override fun copyNew(): Answer {
            return Answer(null, correctAnswers)
        }

        fun toJson(): JSONObject {
            return JSONObject().apply {
                put("user_answer", userAnswer ?: JSONObject.NULL)
                put("correct_answers", JSONArray().apply { correctAnswers.forEach { put(it) } })
            }
        }

        companion object {
            fun fromJson(jsonObject: JSONObject): Answer {
                return Answer(
                    if (jsonObject.isNull("user_answer")) null else jsonObject.getString("user_answer"),
                    jsonObject.getJSONArray("correct_answers").run { List(length()) { i -> getString(i) } }
                )
            }
        }

    }

}